src/Common/CustomerPortalClient.cs

```diff
@@ -22,6 +34,82 @@ public CustomerPortalClient(ISession session)
             _session = session;
         }
 
+        #region Private Methods
+
+        private static DataSet NgpDataSetToDataSet(NgpDataSet ngpDataSet)
+        {
+            DataSet ds = new DataSet();
+
+            if (ngpDataSet == null || (string.IsNullOrWhiteSpace(ngpDataSet.XMLSchema) && string.IsNullOrWhiteSpace(ngpDataSet.DataXML)))
+            {
+                return ds;
+            }
+
+            //Insert schema
+            TextReader a = new StringReader(ngpDataSet.XMLSchema);
+            XmlReader readerS = new XmlTextReader(a);
+            ds.ReadXmlSchema(readerS);
+            XDocument xdS = XDocument.Parse(ngpDataSet.XMLSchema);
+
+            //Insert data
+            UTF8Encoding encoding = new UTF8Encoding();
+            Byte[] byteArray = encoding.GetBytes(ngpDataSet.DataXML);
+            MemoryStream stream = new MemoryStream(byteArray);
+
+            XmlReader reader = new XmlTextReader(stream);
+            try
+            {
+                ds.ReadXml(reader);
+            }
+            catch (ConstraintException ex)
+            {
+                throw new Exception("Error while parsing results from getting other Customer Environments to terminate", ex);
+            }
+            XDocument xd = XDocument.Parse(ngpDataSet.DataXML);
+
+            foreach (DataTable dt in ds.Tables)
+            {
+                var rs = from row in xd.Descendants(dt.TableName)
+                         select row;
+
+                int i = 0;
+                foreach (var r in rs)
+                {
+                    DataRowState state = DataRowState.Added;
+                    if (r.Attribute("RowState") != null)
+                    {
+                        state = (DataRowState)Enum.Parse(typeof(DataRowState), r.Attribute("RowState").Value);
+                    }
+
+                    DataRow dr = dt.Rows[i];
+                    dr.AcceptChanges();
+
+                    switch (state)
+                    {
+                        case DataRowState.Added:
+                            dr.SetAdded();
+                            break;
+                        case DataRowState.Deleted:
+                            dr.Delete();
+                            break;
+                        case DataRowState.Modified:
+                            dr.SetModified();
+                            break;
+                        case DataRowState.Detached:
+                        case DataRowState.Unchanged:
+                        default:
+                            break;
+                    }
+
+                    i++;
+                }
+            }

```

Comment:
Line 72 may be simplified to `var rs = xd.Descendants(dt.TableName);`

Additionally, if we get `rs` as an array and we can convert the `foreach` into a `for` and use the variable `i` to iterate over both `rs` and `dt.Rows`, simplifying the code read.
Something like:
```csharp
foreach (DataTable dt in ds.Tables)
{
    var rs = xd.Descendants(dt.TableName).ToArray();

    for (int i = 0; i < rs.Length; i++)
    {
        var r = rs[i];
        DataRowState state = DataRowState.Added;
        if (r.Attribute("RowState") != null)
        {
            state = (DataRowState)Enum.Parse(typeof(DataRowState), r.Attribute("RowState").Value);
        }

        DataRow dr = dt.Rows[i];
        dr.AcceptChanges();

        switch (state)
        {
            case DataRowState.Added:
                dr.SetAdded();
                break;
            case DataRowState.Deleted:
                dr.Delete();
                break;
            case DataRowState.Modified:
                dr.SetModified();
                break;
            case DataRowState.Detached:
            case DataRowState.Unchanged:
            default:
            break;
        }
    }
}
```

Reply:
that's what happen when you copy utilities that just works :D

Reply:
done